Migrate OpenWeaterMap to new library (support API 3.0)

freekode opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Migration to new library with support of OneCall API (3.0)

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Migrate new library `pyopenweathermap` I created specifically for Home Assistant. Currently, used pyowm library didn't have an update for a long time and seems dead.
Remove forecast modes, one call allows to retrieve daily and hourly forecasts. Remove migration from V1 to V2 of config_flow, forecast mode isn't used anymore, there is no need for migration.

New lib https://github.com/freekode/pyopenweathermap

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: 
  fixes #115714 
  fixes #113000
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

freekode committed:
Migrate integration to new library

freekode committed:
migrate to new forecast

freekode committed:
migrate forecasts responses

freekode committed:
fix test

freekode committed:
upd dependency

freekode committed:
return forecast sensors

freekode committed:
fix forecast cloud sernsor

freekode committed:
remove commented

bdraco commented on homeassistant/components/openweathermap/weather_update_coordinator.py:
```diff
@@ -83,181 +78,122 @@ async def _async_update_data(self):
+                weather_report = await self._owm_client.get_weather(
+                    self._latitude, self._longitude
+                )
+                data = self._convert_weather_response(weather_report)
```
Please move code out of the `try` block that cannot raise the exception being caught

freekode replied:
done, I hope I understood well what you mean

bdraco commented:
Please add a link to the source code of the new library

freekode committed:
move out from try

bdraco commented on homeassistant/components/openweathermap/weather_update_coordinator.py:
```diff
@@ -64,200 +57,142 @@
     async def _async_update_data(self):
         """Update the data."""
-        data = {}
+        weather_report = {}
```
```suggestion
```

Looks like this is unneeded since it with either raise or the value will be set

bdraco commented on homeassistant/components/openweathermap/config_flow.py:
```diff
@@ -54,14 +55,10 @@ async def async_step_user(self, user_input=None):
-                )
-                if not api_online:
+                api_key_valid = await _is_owm_api_key_valid(user_input[CONF_API_KEY])
+                if api_key_valid is False:
                     errors["base"] = "invalid_api_key"
```
```
                if api_key_valid is False:
                    errors["base"] = "invalid_api_key"
```

This block cannot raise any of the exceptions being caught, please move it out of the try

freekode replied:
done

freekode committed:
Update homeassistant/components/openweathermap/weather_update_coordinator.py

Co-authored-by: J. Nick Koston <nick@koston.org>

freekode committed:
move out from try

bdraco commented on homeassistant/components/openweathermap/manifest.json:
```diff
@@ -5,6 +5,6 @@
   "iot_class": "cloud_polling",
-  "loggers": ["geojson", "pyowm", "pysocks"],
-  "requirements": ["pyowm==3.2.0"]
+  "loggers": ["geojson", "pyopenweathermap", "pysocks"],
```
Are these loggers still correct?

bdraco commented on tests/components/openweathermap/test_config_flow.py:
```diff
@@ -28,53 +23,55 @@
-        result = await hass.config_entries.flow.async_init(
-            DOMAIN, context={"source": SOURCE_USER}, data=CONFIG
-        )
+@patch("pyopenweathermap.OWMClient", return_value=_create_mocked_owm_client(True))
```
Usually we patch at the point of import instead of globally

freekode replied:
could you, please, show me an example how it should be done?

freekode replied:
done

freekode committed:
fix tests, remove migration

bdraco commented on homeassistant/components/openweathermap/weather_update_coordinator.py:
```diff
@@ -64,200 +57,141 @@
     async def _async_update_data(self):
         """Update the data."""
-        data = {}
         async with asyncio.timeout(20):
```
Please pass the timeout value to the library and than the aiohttp client request so we don't have two timeouts running

freekode replied:
Added as optional parameter https://github.com/freekode/pyopenweathermap/blob/main/src/pyopenweathermap/owm_client.py#L14
Removed timeout in coordinator

freekode committed:
remove timeout

bdraco commented on tests/components/openweathermap/test_config_flow.py:
```diff
@@ -28,190 +23,155 @@
 
-        assert result["type"] is FlowResultType.FORM
-        assert result["step_id"] == "init"
+@patch("homeassistant.components.openweathermap.config_flow.OWMClient")
```
It probably makes sense to make these pytest fixtures

freekode replied:
updated

freekode committed:
add fixtures

bdraco commented on tests/components/openweathermap/test_config_flow.py:
```diff
@@ -154,22 +171,24 @@ async def test_form_options(
-@patch("homeassistant.components.openweathermap.config_flow.OWMClient")
-async def test_form_api_call_error(owm_client_mock, hass: HomeAssistant) -> None:
+async def test_form_api_call_error(
+    config_flow_owm_client_mock, hass: HomeAssistant
```
Please make `hass` the first fixture for all tests

freekode replied:
done

bdraco commented on homeassistant/components/openweathermap/sensor.py:
```diff
@@ -315,7 +316,9 @@ def __init__(
     def native_value(self) -> StateType:
         """Return the state of the device."""
-        return self._weather_coordinator.data.get(self.entity_description.key, None)
+        return self._weather_coordinator.data[ATTR_API_CURRENT].get(
+            self.entity_description.key, None
+        )
```
```suggestion
        return self._weather_coordinator.data[ATTR_API_CURRENT].get(self.entity_description.key)
```

`None` is the default

`ruff` may complain about the formatting of this suggestion

freekode replied:
done

bdraco commented on homeassistant/components/openweathermap/sensor.py:
```diff
@@ -335,10 +338,7 @@ def __init__(
-            return None
-
+        forecasts = self._weather_coordinator.data[ATTR_API_DAILY_FORECAST]
         value = forecasts[0].get(self.entity_description.key, None)
```
```suggestion
        value = forecasts[0].get(self.entity_description.key)
```

freekode replied:
done

bdraco commented on tests/components/openweathermap/test_config_flow.py:
```diff
@@ -28,190 +24,173 @@
     )
 
-    return mocked_owm
+    assert result["errors"] == {"base": "cannot_connect"}
```
Please finish all config flow tests to reach an abort state or create entry state.

freekode replied:
added test for config_flow abort, test for successful config_flow is already exists

bdraco replied:
Please finish test_form_invalid_api_key and test_form_api_call_error to reach the abort or create_entry state after recovering from the error

freekode replied:
done

bdraco commented:
I think this is close to ready for testing.

I'll give a shot after the next turn

freekode committed:
style fixes

freekode committed:
add abort test

freekode commented:
> I think this is close to ready for testing.
> 
> I'll give a shot after the next turn

FYI you have to have active subscription on OpenWeatherMap. API won't work without it, although OWM won't charge anything.

bdraco commented:
```
Apr 28 21:13:33 homeassistant homeassistant[575]: 2024-04-28 11:13:33.932 DEBUG (MainThread) [homeassistant.config_entries] Config entry 'OpenWeatherMap' for openweathermap integration not ready yet; Retrying in 20 seconds
```

Seems like its raising `UpdateFailed` with an empty string

<img width="788" alt="Screenshot 2024-04-28 at 4 13 52 PM" src="https://github.com/home-assistant/core/assets/663432/19c6dd3c-8c46-4d2e-9f1c-4f19d3d7f842">

freekode commented:
Empty string where?

freekode committed:
upd tests

freekode commented:
> > probably I understand, RequestError has no message
> 
> exactly 👍

Hm..strange, RequestError had message before. I updated lib and added more detailed messages https://github.com/freekode/pyopenweathermap/commit/adadd55f110521fe265ba21ebb4228dc25682748

In the debug I printed out errors, and it seems ok:
![image](https://github.com/home-assistant/core/assets/940893/9b525d8f-3c47-4aed-a419-ee094ce568ca)
![image](https://github.com/home-assistant/core/assets/940893/d4d714da-161f-4d66-b6a3-c4ebf8123e9c)

If you still have issues, let me know what I can change

bdraco commented:
Still failing to setup
```
2024-04-29 08:20:35.582 DEBUG (MainThread) [homeassistant.components.openweathermap.weather_update_coordinator] Finished fetching openweathermap data in 0.385 seconds (success: False)
2024-04-29 08:20:35.582 DEBUG (MainThread) [homeassistant.setup] Phase config_entry_setup for openweathermap (727e03e67745bd15d9923feb667cc0bf) took 0.39s (elapsed=0.39s) (wait_time=0.00s)
```

Not sure whats going on. Will have to dig in some more

bdraco commented:
It looks the error logging was removed in #101483

bdraco commented:
It looks like the error message on the frontend is broken

bdraco commented:
https://github.com/home-assistant/frontend/pull/19128
https://github.com/home-assistant/core/pull/106305

bdraco commented:
The reason is there but its not making it into the UI

freekode committed:
Merge branch 'dev' into owm-upgrade

bdraco commented:
Actual error is

"Please note that using One Call 3.0 requires a separate subscription to the One Call by Call plan. Learn more here https://openweathermap.org/price. If you have a valid subscription to the One Call by Call plan, but still receive this error, then please see https://openweathermap.org/faq#error401 for more info."

Had to dig it out of the api

bdraco commented:
Invalid API key is what I get when I try to set it up again

I've generated 3 keys and still no dice

bdraco commented:
<img width="1059" alt="Screenshot 2024-04-29 at 2 00 30 PM" src="https://github.com/home-assistant/core/assets/663432/9af4bea9-2206-492b-9f94-f2818420ef48">

It says it active

bdraco commented:
It would be nice to get the error back into the UI so the user can figure out whats is going wrong as it seems like activating an api key is not so strait forward

bdraco commented:
example https://github.com/home-assistant/core/blob/c5953045d4c40c3e665efd10066204b05bd89a3d/homeassistant/components/homekit_controller/config_flow.py#L483

bdraco commented:
The error not showing turned out to be a regression in the frontend
https://github.com/home-assistant/frontend/pull/20668

bdraco commented:
I guess we need a repair issue also for users who are upgrading since the 3.0 api requires the subscription  (and apparently doesn't even work if you do get one)


<img width="754" alt="Screenshot 2024-04-29 at 2 14 29 PM" src="https://github.com/home-assistant/core/assets/663432/c6a1c8bb-998b-4c92-9837-14621506abb5">

freekode commented:
Sorry, forgot to mention, they don't activate subscription immediately. I waited several hours.

freekode committed:
Merge branch 'dev' into owm-upgrade

freekode committed:
add description placeholder

freekode committed:
upd lib version

freekode commented:
> example
> 
> https://github.com/home-assistant/core/blob/c5953045d4c40c3e665efd10066204b05bd89a3d/homeassistant/components/homekit_controller/config_flow.py#L483

I added that, but don't know where to check is it working or not

freekode committed:
Merge branch 'dev' into owm-upgrade

Commit:
add error description

freekode commented on homeassistant/components/openweathermap/config_flow.py:
```diff
@@ -53,16 +55,15 @@ async def async_step_user(self, user_input=None):
+                api_key_valid = await _is_owm_api_key_valid(user_input[CONF_API_KEY])
+            except RequestError as error:
                 errors["base"] = "cannot_connect"
+                description_placeholders["error"] = str(error)
```
i added place holder for `cannot_connect` and had to replace translation with eng text.

bdraco commented on homeassistant/components/openweathermap/__init__.py:
```diff
@@ -40,14 +36,11 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        owm_client, latitude, longitude, hass
     )
 
     await weather_coordinator.async_config_entry_first_refresh()
```
We should detect that they don't have the v3 api on their account and create an issue in the issue registry telling them what to do.

https://developers.home-assistant.io/docs/core/platform/repairs?_highlight=issue#creating-an-issue

freekode replied:
i see
i think i can add owm api v2.5 in the lib, then add a switch in HA where you choose which api to use. Then add repair notification to move to 3.0.

bdraco replied:
Sounds good. It should be a config entry option

freekode replied:
ok, i added support v2.5
added migration, all users will be migrated to onecall v2.5
updated config flow, you can choose between v2.5 and v3.0 (default)
if you choose 2.5, during init repair issue will be created

freekode replied:
sorry about mess with emails, i will recreate pr later when everything will be done

freekode replied:
new keys (or rather new accounts) don't work with v2.5 already

freekode replied:
continue here #116870

freekode committed:
v2.5 support

freekode committed:
Merge branch 'dev' into owm-upgrade

freekode committed:
v2.5 support

freekode committed:
Merge branch 'dev' into owm-upgrade

freekode committed:
fix strings

freekode committed:
fix strings

freekode committed:
Merge branch 'dev' into owm-upgrade
